File: backend/src/services/mongo_service.py
"""
MongoDB persistence layer for Secret Santa assignments.
"""
from datetime import datetime, timezone
from typing import List, Optional

from pymongo import MongoClient
from pymongo.errors import ConnectionFailure

from src.models.employee import Assignment, Employee


class MongoService:
    """Handles reading/writing assignments to MongoDB."""

    def __init__(self, uri: str, db_name: str = "secret_santa"):
        self._client = MongoClient(uri, serverSelectionTimeoutMS=5000)
        self._db = self._client[db_name]
        self._assignments = self._db["assignments"]

    def ping(self) -> bool:
        try:
            self._client.admin.command("ping")
            return True
        except ConnectionFailure:
            return False

    def save_assignments(self, assignments: List[Assignment], year: int) -> str:
        """Persist a batch of assignments for a given year. Returns inserted _id."""
        doc = {
            "year": year,
            "created_at": datetime.now(timezone.utc).isoformat(),
            "assignments": [a.to_dict() for a in assignments],
        }
        result = self._assignments.insert_one(doc)
        return str(result.inserted_id)

    def get_assignments_by_year(self, year: int) -> Optional[List[Assignment]]:
        """Retrieve assignments for a given year, or None if not found."""
        doc = self._assignments.find_one({"year": year}, sort=[("created_at", -1)])
        if not doc:
            return None
        return [
            Assignment(
                giver=Employee(name=a["Employee_Name"], email=a["Employee_EmailID"]),
                receiver=Employee(name=a["Secret_Child_Name"], email=a["Secret_Child_EmailID"]),
            )
            for a in doc["assignments"]
        ]

    def list_years(self) -> List[int]:
        """Return all years that have saved assignments."""
        return sorted(self._assignments.distinct("year"))

File: backend-node/src/services/mongo_service.js
const { MongoClient } = require('mongodb');

class MongoService {
  constructor(uri, dbName = 'secret_santa') {
    this.uri = uri;
    this.dbName = dbName;
    this.client = null;
    this.db = null;
    this.assignments = null;
  }

  async connect() {
    try {
      this.client = new MongoClient(this.uri, {
        serverSelectionTimeoutMS: 5000
      });
      await this.client.connect();
      this.db = this.client.db(this.dbName);
      this.assignments = this.db.collection('assignments');
      await this.client.db('admin').command({ ping: 1 });
      return true;
    } catch (err) {
      console.warn(`MongoDB init failed: ${err.message}`);
      return false;
    }
  }

  isConnected() {
    return this.client !== null && this.assignments !== null;
  }

  async saveAssignments(assignments, year) {
    if (!this.isConnected()) {
      throw new Error('MongoDB not connected');
    }

    const doc = {
      year,
      created_at: new Date().toISOString(),
      assignments: assignments.map(a => a.toDict())
    };

    const result = await this.assignments.insertOne(doc);
    return result.insertedId.toString();
  }

  async getAssignmentsByYear(year) {
    if (!this.isConnected()) {
      throw new Error('MongoDB not connected');
    }

    const doc = await this.assignments
      .find({ year })
      .sort({ created_at: -1 })
      .limit(1)
      .toArray();

    if (!doc.length) {
      return null;
    }

    const { Employee, Assignment } = require('../models/employee');
    return doc[0].assignments.map(a => {
      return new Assignment({
        giver: new Employee({ name: a.Employee_Name, email: a.Employee_EmailID }),
        receiver: new Employee({ name: a.Secret_Child_Name, email: a.Secret_Child_EmailID })
      });
    });
  }

  async listYears() {
    if (!this.isConnected()) {
      throw new Error('MongoDB not connected');
    }

    const years = await this.assignments.distinct('year');
    return years.sort((a, b) => a - b);
  }

  async close() {
    if (this.client) {
      await this.client.close();
    }
  }
}

module.exports = { MongoService };
